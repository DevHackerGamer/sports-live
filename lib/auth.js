// Lightweight auth helper that prefers Clerk server verification when available,
// and falls back to a dev header (X-User-Type) if Clerk isn't configured.

let clerkClient = null;
let getAuth = null;
try {
  const clerk = require('@clerk/express');
  clerkClient = clerk.clerkClient;
  getAuth = clerk.getAuth;
} catch (_) {
  // Clerk not installed; dev fallback will be used
}

function getBearer(req) {
  const auth = req.headers['authorization'] || req.headers['Authorization'];
  if (!auth || typeof auth !== 'string') return null;
  const parts = auth.split(' ');
  if (parts.length === 2 && parts[0] === 'Bearer') return parts[1];
  return null;
}

function normalizeRole(meta) {
  if (!meta) return '';
  const type = meta.type || meta.role || '';
  if (Array.isArray(meta.roles) && meta.roles.length) {
    // Prefer explicit admin role if present
    if (meta.roles.map(r => String(r).toLowerCase()).includes('admin')) return 'admin';
  }
  return String(type || '').toLowerCase();
}

// Attempt to resolve user type from Clerk private metadata
async function getUserType(req) {
  // Dev override header
  const devHeaderType = (req.headers['x-user-type'] || '').toString().toLowerCase();
  const devHeaderRole = (req.headers['x-user-role'] || '').toString().toLowerCase();
  const devType = devHeaderType || devHeaderRole;

  // If Clerk middleware/client is not present, return dev
  if (!clerkClient || !getAuth) {
    return devType || '';
  }

  try {
    const { userId } = getAuth(req) || {};
    if (!userId) return devType || '';
  const user = await clerkClient.users.getUser(userId);
  const privateType = normalizeRole(user?.privateMetadata);
  const publicType = normalizeRole(user?.publicMetadata);
  return privateType || publicType || '';
  } catch (e) {
    // On any failure, fall back to dev header
    return devType || '';
  }
}

async function isAdmin(req) {
  const type = await getUserType(req);
  return type === 'admin';
}

module.exports = { getUserType, isAdmin };
