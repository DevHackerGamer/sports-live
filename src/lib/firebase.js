// Firebase client initialization (App + Realtime Database)
import { initializeApp, getApps, getApp } from 'firebase/app';
import { getAnalytics } from 'firebase/analytics';
import {
  getDatabase,
  ref as dbRef,
  update as dbUpdate,
  set as dbSet,
  child as dbChild,
  get as dbGet,
  onValue as dbOnValue  // <-- add this
} from 'firebase/database';

// Firebase web config (public by design)
const firebaseConfig = {
  apiKey: 'AIzaSyDzZ0CAodwCYHgQJahsgcxJrmWiIIbVGM',
  authDomain: 'sports-live-c25ca.firebaseapp.com',
  databaseURL: 'https://sports-live-c25ca-default-rtdb.firebaseio.com',
  projectId: 'sports-live-c25ca',
  storageBucket: 'sports-live-c25ca.firebasestorage.app',
  messagingSenderId: '532399482994',
  appId: '1:532399482994:web:3d1353393886b64dc1baeb',
  measurementId: 'G-8JT6X5DF6B',
};

// Singleton app instance across HMR/dev reloads
const app = getApps().length ? getApp() : initializeApp(firebaseConfig);

// Realtime Database instance
export const db = getDatabase(app);

// Modular functions for convenience
export const ref = dbRef;
export const update = dbUpdate;
export const set = dbSet;
export const child = dbChild;
export const get = dbGet;
export const onValue = dbOnValue;  // <-- export onValue

// Guard analytics for browser-only environments
export const initAnalytics = () => {
  if (typeof window !== 'undefined' && 'measurementId' in firebaseConfig) {
    try {
      return getAnalytics(app);
    } catch (_) {
      // ignore if analytics not supported
    }
  }
  return null;
};

export default app;
