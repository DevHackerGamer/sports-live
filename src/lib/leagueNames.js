// Centralized league name mapping for ESPN league codes
// This provides consistent display names across the entire application

export const LEAGUE_NAMES = {
  // ESPN league codes
  'eng.1': 'Premier League',
  'esp.1': 'La Liga',
  'ita.1': 'Serie A',
  'ger.1': 'Bundesliga',
  'fra.1': 'Ligue 1',
  'uefa.champions': 'Champions League',
  'ned.1': 'Eredivisie',
  'por.1': 'Primeira Liga',
  'eng.2': 'Championship',
  'bra.1': 'Campeonato Brasileiro SÃ©rie A',
  'conmebol.libertadores': 'Copa Libertadores',
  
  // Football-data.org codes (legacy)
  'PL': 'Premier League',
  'PD': 'La Liga',
  'SA': 'Serie A',
  'BL1': 'Bundesliga',
  'FL1': 'Ligue 1',
  'CL': 'Champions League',
  'DED': 'Eredivisie',
  'PPL': 'Primeira Liga',
  'ELC': 'Championship'
};

/**
 * Get display name for a league code
 * @param {string} code - ESPN league code (e.g., 'eng.1') or name with bracket notation (e.g., 'Bundesliga [ger.1]')
 * @returns {string} Display name (e.g., 'Premier League')
 */
export function getLeagueName(code) {
  if (!code) return 'Unknown League';
  
  // Handle strings with bracket notation like "Bundesliga [ger.1]"
  const bracketMatch = /^(.*?)\s*\[([^\]]+)\]$/.exec(String(code));
  if (bracketMatch) {
    const [, namePart, codePart] = bracketMatch;
    // Try to get the name from the code part first
    const nameFromCode = LEAGUE_NAMES[codePart];
    if (nameFromCode) return nameFromCode;
    // Otherwise return the name part without brackets
    return namePart.trim();
  }
  
  return LEAGUE_NAMES[code] || code;
}

/**
 * Get league code from display name or return as-is if already a code
 * @param {string} nameOrCode - League name or code
 * @returns {string} League code
 */
export function getLeagueCode(nameOrCode) {
  if (!nameOrCode) return '';
  
  // If it's already a code, return it
  if (LEAGUE_NAMES[nameOrCode]) return nameOrCode;
  
  // Find code by name
  const entry = Object.entries(LEAGUE_NAMES).find(([_, name]) => name === nameOrCode);
  return entry ? entry[0] : nameOrCode;
}

/**
 * Format competition object to ensure it has proper display name
 * @param {Object} competition - Competition object with name/code
 * @returns {Object} Competition with displayName
 */
export function formatCompetition(competition) {
  if (!competition) return { name: 'Unknown League', displayName: 'Unknown League' };
  
  const name = competition.name || competition.code || competition.id || '';
  const displayName = getLeagueName(name);
  
  return {
    ...competition,
    displayName,
    code: competition.code || name
  };
}
