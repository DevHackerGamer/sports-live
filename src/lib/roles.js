// Client-side role utilities to interpret Clerk metadata consistently

export function getUserRoles(user) {
  if (!user) return [];
  const raw = [
    user?.privateMetadata?.type,
    user?.publicMetadata?.type,

  ]
    .concat(Array.isArray(user?.privateMetadata?.roles) ? user.privateMetadata.roles : [])
    .concat(Array.isArray(user?.publicMetadata?.roles) ? user.publicMetadata.roles : [])
    .filter(Boolean)
    .map(v => String(v).toLowerCase());
  return Array.from(new Set(raw));
}

export function isAdminFromUser(user) {
  const roles = getUserRoles(user);
  return roles.includes('admin');
}
