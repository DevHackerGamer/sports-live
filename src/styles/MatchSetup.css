.match-setup {
  max-width: 1000px;
  margin: 0 auto;
  padding: 1rem;
}

.header-row {
  display: flex;
  justify-content: space-between;
  align-items: center;
  margin-bottom: 2rem;
  padding-bottom: 1rem;
  border-bottom: 1px solid #e5e5e5;
}

.header-row h2 {
  margin: 0;
  font-size: 1.8rem;
  font-weight: 300;
  color: #1a1a1a;
}

.btn-primary {
  background: #1a1a1a;
  color: white;
  border: none;
  padding: 12px 24px;
  border-radius: 8px;
  font-weight: 400;
  cursor: pointer;
  transition: all 0.2s ease;
  font-size: 1rem;
}

.btn-primary:hover {
  background: #2d2d2d;
  transform: translateY(-1px);
}

.btn-danger {
  background: #e74c3c;
  color: white;
  border: none;
  padding: 8px 16px;
  border-radius: 6px;
  font-weight: 400;
  cursor: pointer;
  transition: all 0.2s ease;
  font-size: 0.9rem;
}

.btn-danger:hover {
  background: #c0392b;
  transform: translateY(-1px);
}

.match-form {
  background: white;
  padding: 2rem;
  border-radius: 8px;
  box-shadow: 0 2px 4px rgba(0,0,0,0.02);
  border: 1px solid #e5e5e5;
  margin-bottom: 2.5rem;
}

.match-form h3 {
  margin-top: 0;
  margin-bottom: 1.5rem;
  font-size: 1.4rem;
  font-weight: 400;
  color: #1a1a1a;
}

.form-row {
  display: flex;
  gap: 1rem;
  margin-bottom: 1.5rem;
  align-items: center;
}

.form-row input {
  flex: 1;
  padding: 0.75rem;
  border: 1px solid #e5e5e5;
  border-radius: 4px;
  font-size: 1rem;
  transition: border-color 0.2s ease;
}

.form-row input:focus {
  outline: none;
  border-color: #1a1a1a;
}

.form-row .vs {
  font-weight: 500;
  color: #666;
  margin: 0 0.5rem;
}

.matches-list h3 {
  margin-bottom: 1.5rem;
  font-size: 1.4rem;
  font-weight: 400;
  color: #1a1a1a;
  padding-bottom: 0.5rem;
  border-bottom: 1px solid #e5e5e5;
}

.match-item {
  display: flex;
  justify-content: space-between;
  align-items: center;
  background: #f8f8f8;
  padding: 1.5rem;
  border-radius: 8px;
  border: 1px solid #e5e5e5;
  margin-bottom: 1.5rem;
  transition: all 0.2s ease;
}

.match-item:hover {
  transform: translateY(-1px);
  box-shadow: 0 4px 12px rgba(0, 0, 0, 0.1);
}

.match-info {
  flex: 1;
}

.match-info .teams {
  display: flex;
  align-items: center;
  margin-bottom: 0.75rem;
  font-size: 1.1rem;
}

.match-info .team {
  font-weight: 500;
  color: #1a1a1a;
}

.match-info .vs {
  margin: 0 0.75rem;
  color: #666;
  font-weight: 400;
}

.match-details {
  display: flex;
  gap: 1rem;
  font-size: 0.9rem;
  color: #666;
}

.match-details span {
  background: #f0f0f0;
  padding: 4px 8px;
  border-radius: 4px;
}

.no-matches {
  text-align: center;
  padding: 3rem;
  background: white;
  border-radius: 8px;
  box-shadow: 0 2px 4px rgba(0,0,0,0.02);
  border: 1px solid #e5e5e5;
  color: #7f8c8d;
}

.no-matches p {
  margin: 0;
  font-size: 1.1rem;
}

/* Responsive design */
@media (max-width: 768px) {
  .match-setup {
    padding: 0.5rem;
  }
  
  .header-row {
    flex-direction: column;
    align-items: flex-start;
    gap: 1rem;
  }
  
  .form-row {
    flex-direction: column;
    gap: 0.75rem;
  }
  
  .form-row .vs {
    margin: 0;
    text-align: center;
  }
  
  .match-item {
    flex-direction: column;
    align-items: flex-start;
    gap: 1rem;
  }
  
  .match-details {
    flex-wrap: wrap;
  }
}