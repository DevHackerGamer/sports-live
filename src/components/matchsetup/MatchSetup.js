import React, { useEffect, useState } from 'react';
import { useUser } from '@clerk/clerk-react';
import MatchViewer from '../matchViewer/MatchViewer';
import { isAdminFromUser } from '../../lib/roles';
import '../../styles/MatchSetup.css';

//Component for setting up and managing matches
// Admin-only screen for creating/scheduling matches.
const MatchSetup = ({ isAdmin: isAdminProp }) => {
  const { user } = useUser();
  const [matches, setMatches] = useState([]);
  const [teams,setTeams] = useState([]);
  const [newMatch, setNewMatch] = useState({
    teamA: {},
    teamB: {},
    date: '',
    time: '',
    competition: ''
  });
  const [showForm, setShowForm] = useState(false);
  const [loading,setLoading] = useState(false);
  const [selectedMatch, setSelectedMatch] = useState(null);

  const isAdmin = typeof isAdminProp === 'boolean' ? isAdminProp : isAdminFromUser(user);

  // fetch only admin-created matches using createdMatchs api
  useEffect(() => {
    const fetchMatches = async () => {
      try {
        const res = await fetch('/api/createdMatches');
        const matchData = await res.json();

        // Only upcoming matches
        const now = new Date();
        const upcomingMatches = [];

        for (let match of matchData) {
          const matchDate = new Date(`${match.date}T${match.time}`);
          if (matchDate >= now) {
            upcomingMatches.push(match);
          } else {
            // delete past matches automatically to scale db
            await fetch(`/api/createdMatches?id=${match._id}`, { method: "DELETE" });
          }
        }

        setMatches(upcomingMatches);
      } catch (err) {
        console.error("Error fetching admin-created matches:", err);
      }
    };

    fetchMatches();
  }, []);

  // fetch all teams for the dropdown input data also ensure valid teams are selected
  useEffect(() => {
    const fetchTeams = async () => {
      try {
        const res = await fetch('/api/teams');
        if (!res.ok) throw new Error('Failed to fetch teams');
        const json = await res.json();
        setTeams(json.data.map(t => ({
          id: t._id || t.id,
          name: typeof t.name === 'string' ? t.name : t.name?.en || 'Unnamed Team'
        })));
      } catch (err) {
        console.error('Error fetching teams:', err);
      }
    };

    fetchTeams();
  }, []);

  if (!isAdmin) {
    return (
      <div className="match-setup">
        <h2>Match Setup</h2>
        <p style={{ color: '#b00' }}>Access denied: Admin role required.</p>
      </div>
    );
  }

  //Handle input changes
  const handleInputChange = (e) => {
    const { name, value } = e.target;
    setNewMatch({ ...newMatch, [name]: value });
  };

  // match validation
  const validateMatch = () => {
    if(!newMatch.teamA.id || !newMatch.teamB.id || !newMatch.date || !newMatch.time){
      alert('Fill in all required fields');
      return false;
    }
    if(newMatch.teamA.id === newMatch.teamB.id){
      alert('Home and Away teams must be different');
      return false;
    }
    return true;
  }

  //Add match to the AdminCreatedMatchs collection in db
  const addMatch = async () => {
    if(!validateMatch()) return;
    setLoading(true);
    try {
      const res = await fetch('/api/createdMatches', {
        method: 'POST',
        headers: {'Content-type':'application/json'},
        body: JSON.stringify({
          homeTeam: newMatch.teamA,
          awayTeam: newMatch.teamB,
          date: newMatch.date,
          time: newMatch.time,
          competition: { id: 'unknown', name: { en: newMatch.competition || 'Unknown Competition' } }
        })
      });
      if(!res.ok){
        const error = await res.json();
        throw new Error(error.error || 'Failed to save match');
      } else {
        const matchRes = await fetch('/api/createdMatches');
        const updatedMatches = await matchRes.json();
        setMatches(updatedMatches);
        setNewMatch({teamA:{},teamB:{},date:'',time:'',competition:''});
        setShowForm(false);
      }
    } catch(error){
      console.error('Error saving match',error);
      alert('Failed to save match, check console for details');
    } finally {
      setLoading(false);
    }
  }

  //Remove match from the AdminCreatedMatches collection in the db
  const removeMatch = async(id) => {
    if(!window.confirm('Are you sure you want to remove this match?')) return;
    try {
      const res = await fetch(`/api/createdMatches?id=${id}`, { method:'DELETE' });
      if(!res.ok){
        const error = await res.json();
        throw new Error(error.error || 'Failed to delete match');
      }
      setMatches(matches.filter(m => m._id !== id));
    } catch(error){
      console.error('Error deleting match', error);
      alert('Could not delete match, see console for details');
    }
  };

  // Handle selecting a match to view details
  const handleMatchSelect = (match) => {
    setSelectedMatch(match);
  };

  return (
    <div className="match-setup">
      {selectedMatch ? (
        <MatchViewer 
          match={selectedMatch} 
          initialSection="details" 
          onBack={() => setSelectedMatch(null)}
        />
      ) : (
        <>
          <div className="header-row">
            <h2>Match Setup</h2>
            <button className="btn-primary" onClick={() => setShowForm(!showForm)}>
              {showForm ? 'Cancel' : 'Create New Match'}
            </button>
          </div>

          {showForm && (
            <div className="match-form">
              <h3>Create New Match</h3>
              <div className="form-row">
                <select
                  name="teamA"
                  value={newMatch.teamA.id || ''}
                  onChange={e => {
                    const team = teams.find(t => t.id === e.target.value);
                    setNewMatch(prev => ({ ...prev, teamA: team || {} }));
                  }}
                >
                  <option value="">Select Home Team</option>
                  {teams.map(team => (
                    <option key={team.id} value={team.id}>
                      {team.name}
                    </option>
                  ))}
                </select>

                <span className="vs">VS</span>

                <select
                  name="teamB"
                  value={newMatch.teamB.id || ''}
                  onChange={e => {
                    const team = teams.find(t => t.id === e.target.value);
                    setNewMatch(prev => ({ ...prev, teamB: team || {} }));
                  }}
                >
                  <option value="">Select Away Team</option>
                  {teams.map(team => (
                    <option key={team.id} value={team.id}>
                      {team.name}
                    </option>
                  ))}
                </select>
              </div>
              <div className="form-row">
                <input type="date" name="date" value={newMatch.date} onChange={handleInputChange}/>
                <input type="time" name="time" value={newMatch.time} onChange={handleInputChange}/>
              </div>
              <div className="form-row">
                <input type="text" name="competition" placeholder="Competition (e.g., Premier League)" value={newMatch.competition} onChange={handleInputChange}/>
              </div>
              <button className="btn-primary" onClick={addMatch} disabled={loading}>
                {loading ? 'Saving...' : 'Create Match'}
              </button>
            </div>
          )}

          <div className="matches-list">
            <h3>Scheduled Matches</h3>
            {matches.length === 0 ? (
              <p className="no-matches">No matches scheduled yet.</p>
            ) : (
              matches.map(match => (
                <div 
                  key={match._id} 
                  className="match-item"
                  style={{ cursor: 'pointer' }}
                  onClick={() => handleMatchSelect(match)}
                >
                  <div className="match-info">
                    <div className="teams">
                      <span className="team">{match.homeTeam?.name?.en || match.homeTeam?.name || 'Unnamed Team'}</span>
                      <span className="vs">vs</span>
                      <span className="team">{match.awayTeam?.name?.en || match.awayTeam?.name || 'Unnamed Team'}</span>
                    </div>
                    <div className="match-details">
                      <span className="competition">{match.competition?.name?.en || match.competitionName || 'Unknown Competition'}</span>
                      <span className="datetime">{match.date} at {match.time}</span>
                    </div>
                  </div>
                  <button 
                    className="btn-danger" 
                    onClick={(e) => { e.stopPropagation(); removeMatch(match._id); }}
                  >
                    Remove
                  </button>
                </div>
              ))
            )}
          </div>
        </>
      )}
    </div>
  );
};

export default MatchSetup;