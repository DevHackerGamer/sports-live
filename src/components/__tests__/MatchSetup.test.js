import React from 'react';
import { render, screen, fireEvent, waitFor, act } from '@testing-library/react';
import MatchSetup from '../matchsetup/MatchSetup';
import * as ClerkReact from '@clerk/clerk-react';
import * as roles from '../../lib/roles';

// Mock Clerk
jest.mock('@clerk/clerk-react', () => ({
  useUser: jest.fn(),
}));

// Mock child components
jest.mock('../matchViewer/MatchViewer', () => ({ match, onBack, isAdmin }) => (
  <div data-testid="match-viewer">
    MatchViewer: {match.homeTeam?.name} vs {match.awayTeam?.name}
    <button onClick={onBack}>Back to Setup</button>
  </div>
));

// Mock utilities
jest.mock('../../lib/roles', () => ({
  isAdminFromUser: jest.fn(),
}));

jest.mock('../../lib/leagueNames', () => ({
  getLeagueName: jest.fn((name) => name || 'Unknown League'),
}));

describe('MatchSetup Component', () => {
  const mockOnTeamSelect = jest.fn();
  const mockOnMatchSelect = jest.fn();
  const mockUser = { id: 'user123', firstName: 'John' };

  const mockTeamsData = {
    data: [
      { _id: 'team1', name: 'Arsenal', crest: '/arsenal.png' },
      { _id: 'team2', name: 'Chelsea', crest: '/chelsea.png' },
      { _id: 'team3', name: 'Liverpool', crest: '/liverpool.png' },
    ],
  };

  const mockMatchesData = {
    success: true,
    provider: 'admin',
    data: [
      {
        id: 'match1',
        homeTeam: { name: 'Arsenal', id: 'team1' },
        awayTeam: { name: 'Chelsea', id: 'team2' },
        competition: { name: 'Premier League [eng.1]' },
        utcDate: new Date(Date.now() + 24 * 60 * 60 * 1000).toISOString(),
        date: new Date(Date.now() + 24 * 60 * 60 * 1000).toISOString().split('T')[0],
        time: '15:00',
        status: 'TIMED',
        matchday: 1,
        createdByAdmin: true,
      },
    ],
  };

  const mockCompetitionsData = {
    success: true,
    data: ['Premier League [eng.1]', 'La Liga [esp.1]', 'Champions League [uefa.champions]'],
  };

  beforeEach(() => {
    jest.clearAllMocks();
    
    // Default mocks
    ClerkReact.useUser.mockReturnValue({ user: mockUser });
    jest.spyOn(roles, 'isAdminFromUser').mockReturnValue(true);
    
    // Mock fetch responses
    global.fetch = jest.fn();
    fetch.mockImplementation((url) => {
      if (url.includes('/api/teams')) {
        return Promise.resolve({
          ok: true,
          json: () => Promise.resolve(mockTeamsData),
        });
      }
      if (url.includes('/api/matches?provider=admin')) {
        return Promise.resolve({
          ok: true,
          json: () => Promise.resolve(mockMatchesData),
        });
      }
      if (url.includes('/api/competitions')) {
        return Promise.resolve({
          ok: true,
          json: () => Promise.resolve(mockCompetitionsData),
        });
      }
      if (url.includes('/api/matches')) {
        return Promise.resolve({
          ok: true,
          json: () => Promise.resolve({
            success: true,
            data: {
              id: 'new-match',
              homeTeam: { name: 'Arsenal', id: 'team1' },
              awayTeam: { name: 'Chelsea', id: 'team2' },
              competition: { name: 'Premier League' },
              utcDate: new Date().toISOString(),
              date: new Date().toISOString().split('T')[0],
              time: '15:00',
              status: 'TIMED',
              matchday: 1,
              createdByAdmin: true,
            },
          }),
        });
      }
      return Promise.resolve({
        ok: true,
        json: () => Promise.resolve({}),
      });
    });
  });

  it('renders access denied for non-admin users', () => {
    jest.spyOn(roles, 'isAdminFromUser').mockReturnValue(false);

    render(<MatchSetup onTeamSelect={mockOnTeamSelect} onMatchSelect={mockOnMatchSelect} />);

    expect(screen.getByText('Access denied: Admin role required.')).toBeInTheDocument();
  });

  it('renders main interface for admin users', async () => {
    await act(async () => {
      render(<MatchSetup onTeamSelect={mockOnTeamSelect} onMatchSelect={mockOnMatchSelect} />);
    });

    expect(screen.getByText('Match Setup')).toBeInTheDocument();
    expect(screen.getByText('Create Match')).toBeInTheDocument();
    expect(screen.getByText('Scheduled Matches')).toBeInTheDocument();
  });

  it('fetches teams, matches, and competitions on mount', async () => {
    await act(async () => {
      render(<MatchSetup onTeamSelect={mockOnTeamSelect} onMatchSelect={mockOnMatchSelect} />);
    });

    await waitFor(() => {
      expect(fetch).toHaveBeenCalledWith('/api/teams');
      expect(fetch).toHaveBeenCalledWith('/api/matches?provider=admin&limit=50&range=30&includePast=1');
      expect(fetch).toHaveBeenCalledWith('/api/competitions');
    });
  });

  it('shows and hides the create match form', async () => {
    await act(async () => {
      render(<MatchSetup onTeamSelect={mockOnTeamSelect} onMatchSelect={mockOnMatchSelect} />);
    });

    // Form should be hidden initially
    expect(screen.queryByText('Create New Match')).not.toBeInTheDocument();

    // Show form
    const createButton = screen.getByText('Create Match');
    fireEvent.click(createButton);

    expect(screen.getByText('Create New Match')).toBeInTheDocument();
    expect(screen.getByText('Close Form')).toBeInTheDocument();

    // Hide form
    fireEvent.click(screen.getByText('Close Form'));

    expect(screen.queryByText('Create New Match')).not.toBeInTheDocument();
  });

  it('validates form before submitting', async () => {
    await act(async () => {
      render(<MatchSetup onTeamSelect={mockOnTeamSelect} onMatchSelect={mockOnMatchSelect} />);
    });

    // Show form
    fireEvent.click(screen.getByText('Create Match'));

    // Try to submit empty form
    const submitButton = screen.getByText('Create Match');
    fireEvent.click(submitButton);

    // Should show validation errors
    await waitFor(() => {
      expect(screen.getByText('Home team is required')).toBeInTheDocument();
      expect(screen.getByText('Away team is required')).toBeInTheDocument();
      expect(screen.getByText('Date is required')).toBeInTheDocument();
      expect(screen.getByText('Time is required')).toBeInTheDocument();
      expect(screen.getByText('Competition is required')).toBeInTheDocument();
    });
  });

  
  it('handles team selection from match cards', async () => {
    await act(async () => {
      render(<MatchSetup onTeamSelect={mockOnTeamSelect} onMatchSelect={mockOnMatchSelect} />);
    });

    await waitFor(() => {
      expect(screen.getByText('Arsenal')).toBeInTheDocument();
    });

    const teamCrests = screen.getAllByAltText(/crest/);
    fireEvent.click(teamCrests[0]);

    expect(mockOnTeamSelect).toHaveBeenCalledWith(expect.objectContaining({
      name: 'Arsenal',
      crest: expect.any(String),
    }));
  });

  it('handles match selection', async () => {
    await act(async () => {
      render(<MatchSetup onTeamSelect={mockOnTeamSelect} onMatchSelect={mockOnMatchSelect} />);
    });

    await waitFor(() => {
      expect(screen.getByText('Arsenal')).toBeInTheDocument();
    });

    const matchCard = screen.getByText('Arsenal').closest('.ms-match-card');
    fireEvent.click(matchCard);

    expect(mockOnMatchSelect).toHaveBeenCalledWith(mockMatchesData.data[0]);
  });

  it('shows match viewer when match is selected', async () => {
    await act(async () => {
      render(<MatchSetup onTeamSelect={mockOnTeamSelect} onMatchSelect={mockOnMatchSelect} />);
    });

    await waitFor(() => {
      expect(screen.getByText('Arsenal')).toBeInTheDocument();
    });

    const matchCard = screen.getByText('Arsenal').closest('.ms-match-card');
    fireEvent.click(matchCard);

    expect(screen.getByTestId('match-viewer')).toBeInTheDocument();
    expect(screen.getByText('Back to Setup')).toBeInTheDocument();
  });


  it('cancels remove confirmation', async () => {
    await act(async () => {
      render(<MatchSetup onTeamSelect={mockOnTeamSelect} onMatchSelect={mockOnMatchSelect} />);
    });

    await waitFor(() => {
      expect(screen.getByText('Arsenal')).toBeInTheDocument();
    });

    const removeButtons = screen.getAllByText('Remove');
    fireEvent.click(removeButtons[0]);

    const cancelButton = screen.getByText('Cancel');
    fireEvent.click(cancelButton);

    expect(screen.queryByText('Remove Match')).not.toBeInTheDocument();
  });

  

  it('shows loading state while fetching matches', async () => {
    // Delay the matches response
    fetch.mockImplementation((url) => {
      if (url.includes('/api/matches?provider=admin')) {
        return new Promise(resolve => {
          setTimeout(() => resolve({
            ok: true,
            json: () => Promise.resolve(mockMatchesData),
          }), 100);
        });
      }
      return Promise.resolve({
        ok: true,
        json: () => Promise.resolve({}),
      });
    });

    render(<MatchSetup onTeamSelect={mockOnTeamSelect} onMatchSelect={mockOnMatchSelect} />);

    expect(screen.getByText('Loading matches...')).toBeInTheDocument();

    await waitFor(() => {
      expect(screen.queryByText('Loading matches...')).not.toBeInTheDocument();
    });
  });

  it('shows empty state when no matches exist', async () => {
    fetch.mockImplementation((url) => {
      if (url.includes('/api/matches?provider=admin')) {
        return Promise.resolve({
          ok: true,
          json: () => Promise.resolve({ success: true, data: [] }),
        });
      }
      return Promise.resolve({
        ok: true,
        json: () => Promise.resolve({}),
      });
    });

    await act(async () => {
      render(<MatchSetup onTeamSelect={mockOnTeamSelect} onMatchSelect={mockOnMatchSelect} />);
    });

    expect(screen.getByText('No matches scheduled yet.')).toBeInTheDocument();
  });

  it('handles API errors gracefully', async () => {
    fetch.mockImplementation((url) => {
      if (url.includes('/api/teams')) {
        return Promise.reject(new Error('Failed to fetch teams'));
      }
      return Promise.resolve({
        ok: true,
        json: () => Promise.resolve({}),
      });
    });

    await act(async () => {
      render(<MatchSetup onTeamSelect={mockOnTeamSelect} onMatchSelect={mockOnMatchSelect} />);
    });

    // Should still render without crashing
    expect(screen.getByText('Match Setup')).toBeInTheDocument();
  });

  it('validates matchday input', async () => {
    await act(async () => {
      render(<MatchSetup onTeamSelect={mockOnTeamSelect} onMatchSelect={mockOnMatchSelect} />);
    });

    fireEvent.click(screen.getByText('Create Match'));

    const matchdayInput = screen.getByPlaceholderText('Matchday (optional)');
    fireEvent.change(matchdayInput, { target: { value: '0' } });

    const submitButton = screen.getByText('Create Match');
    fireEvent.click(submitButton);

    await waitFor(() => {
      expect(screen.getByText('Matchday must be a number greater than 0')).toBeInTheDocument();
    });
  });

  it('validates team uniqueness', async () => {
    await act(async () => {
      render(<MatchSetup onTeamSelect={mockOnTeamSelect} onMatchSelect={mockOnMatchSelect} />);
    });

    fireEvent.click(screen.getByText('Create Match'));

    const homeTeamSelect = screen.getByDisplayValue('Select Home Team');
    const awayTeamSelect = screen.getByDisplayValue('Select Away Team');

    // Select same team for both home and away
    fireEvent.change(homeTeamSelect, { target: { value: 'team1' } });
    fireEvent.change(awayTeamSelect, { target: { value: 'team1' } });

    const submitButton = screen.getByText('Create Match');
    fireEvent.click(submitButton);

    await waitFor(() => {
      expect(screen.getByText('Home and Away teams must be different')).toBeInTheDocument();
    });
  });



  it('matches snapshot with admin access', async () => {
    await act(async () => {
      const { container } = render(<MatchSetup onTeamSelect={mockOnTeamSelect} onMatchSelect={mockOnMatchSelect} />);
      expect(container).toMatchSnapshot();
    });
  });

  it('matches snapshot with non-admin access', () => {
    jest.spyOn(roles, 'isAdminFromUser').mockReturnValue(false);

    const { container } = render(<MatchSetup onTeamSelect={mockOnTeamSelect} onMatchSelect={mockOnMatchSelect} />);
    expect(container).toMatchSnapshot();
  });



  
  
});