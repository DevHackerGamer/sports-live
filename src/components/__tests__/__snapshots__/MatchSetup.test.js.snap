// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MatchSetup Component matches snapshot with admin access 1`] = `<div />`;

exports[`MatchSetup Component matches snapshot with non-admin access 1`] = `
<div>
  <div
    class="ms-container"
  >
    <h2>
      Match Setup
    </h2>
    <p
      style="color: rgb(187, 0, 0);"
    >
      Access denied: Admin role required.
    </p>
  </div>
</div>
`;
